recommence/Checkpoint.py

```diff
@@ -58,6 +63,14 @@ def save(self) -> None:
             target=self._c.save_path,
         )
 
+    def _save_every(self) -> None:
+        def _save_every_thread(self) -> None:
+            while True:
+                self.save()
```

Comment:
We'll want to do some sort of locking here to prevent the main thread from moving forward while we are mid-save. We want to avoid the possibility of an inconsistent save object, e.g.:
```python
state = { 'a': 1, 'b': 2 }
start_saving(state)
state['b'] = 3

loaded = load_saved_state()
print(loaded) # -> { 'a': 1, 'b': 3 }
```
where `b` was modified after calling "save" but before actually being written to disk.

I _think_ the GIL will probably make this a safe operation, but we can guarantee this consistency using a mutex lock.
```python
from threading import Lock

def __init__(self):
  self._lock = Lock()

def _save_every(self):
  ...
  def _save_every_thread():
    ...
    with self._lock:
      self.save()
```

The major challenge, though, is that the main thread needs some way to check whether it currently holds the lock or not. This does make me wonder if we ultimately will need a `maybe_save()` method that the caller needs to stick within their main loop

Reply:
I did not know about this issue. I will be doing it in the simpler way using `maybe_save` in the main loop. I need to ask you how the lock works, and learn how to implement this version later